### PR TITLE
James 1784 - Add authentication when downloading an attachment

### PR DESCRIPTION
**Pull request by aduprat**
Rebased #303 other #305

### CODE REVIEW

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/DownloadStepdefs.java [hunk 1]
```diff
@@ -92,26 +103,78 @@ public void downloads(String username, String attachmentId) throws Throwable {
...
 
-    @When("^\"([^\"]*)\" asks for an attachment without blobId parameter$")
-    public void getDownloadWithoutBlobId(String username) throws Throwable {
+    @When("^\"([^\"]*)\" downloads \"([^\"]*)\" without blobId parameter$")
+    public void getDownloadWithoutBlobId(String username, String attachmentId) throws Throwable {
+        String blobId = blobIdByAttachmentId.get(attachmentId);
         AccessToken accessToken = userStepdefs.tokenByUser.get(username);
         response = with()
-            .header("Authorization", accessToken.serialize())
+            .param("access_token", attachmentAccessToken(blobId, accessToken))
```
Comment (rouazana) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/DownloadStepdefs.java]: Did you really replace actual tests with Authorization headers with access_token param?
I think we should still support Authorization headers

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/DownloadEndpoint.feature [hunk 2]
```diff
@@ -17,35 +17,58 @@ Feature: Download endpoint
...
+    Given "usera@domain.tld" is connected
+    When "usera@domain.tld" downloads "a1" with an invalid authentication token
+    Then the user should not be authorized
+
   Scenario: An authenticated user should have access to the download endpoint
     Given "usera@domain.tld" is connected
     When "usera@domain.tld" downloads "a1"
     Then the user should be authorized
 
-  @Ignore
```
Comment (rouazana) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/DownloadEndpoint.feature]: I think we should keep these tests are reactivate them when Inbox will be ready

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/DownloadEndpoint.feature [hunk 3]
```diff
@@ -17,35 +17,58 @@ Feature: Download endpoint
     When "usera@domain.tld" checks for the availability of the attachment endpoint
     Then the user should be authorized
 
+  Scenario: An unauthenticated user should not have access to the download endpoint
+    When "usera@domain.tld" downloads "a1"
+    Then the user should not be authorized
```
Comment (rouazana) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/DownloadEndpoint.feature]: So we are breaking Inbox here?

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/DownloadEndpoint.feature [hunk 4]
```diff
@@ -17,35 +17,58 @@ Feature: Download endpoint
...
+
+  Scenario: An authenticated user should not have access to the download endpoint without the authentication token
+    Given "usera@domain.tld" is connected
+    When "usera@domain.tld" downloads "a1" without the authentication token
+    Then the user should not be authorized
```
Comment (rouazana) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/DownloadEndpoint.feature]: Seems wrong, if connected it should have an Authorization token and then should be able to download

Comment (rouazana) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/DownloadEndpoint.feature]: Idem in the followin tests?

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/DownloadEndpoint.feature [hunk 5]
```diff
@@ -17,35 +17,58 @@ Feature: Download endpoint
...
+
+  Scenario: An authenticated user should not have access to the download endpoint with an empty authentication token
+    Given "usera@domain.tld" is connected
```
Comment (rouazana) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/DownloadEndpoint.feature]: Do you really need this given as this is a connection less GET?
(and in other tests)

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/DownloadStepdefs.java (hunk 1 above)
Comment (aduprat) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/DownloadStepdefs.java]: Yes, we are only supporting this way of authentication now.

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/DownloadEndpoint.feature (hunk 2 above)
Comment (aduprat) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/DownloadEndpoint.feature]: This test is already covered (line 20)

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/DownloadEndpoint.feature (hunk 3 above)
Comment (aduprat) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/DownloadEndpoint.feature]: That's it

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/DownloadStepdefs.java (hunk 1 above)
Comment (rouazana) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/DownloadStepdefs.java]: We have already written the tests to support the main way. I don't think it is a good idea to remove or replace these tests.

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/DownloadEndpoint.feature (hunk 4 above)
Comment (aduprat) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/DownloadEndpoint.feature]: `authenticated` means that the user has already access to the POST method, then he is trying to access the GET method.

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/DownloadEndpoint.feature (hunk 5 above)
Comment (aduprat) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/DownloadEndpoint.feature]: Same response as in previous comment.

**Comment by aduprat** [comment]
#308

**Pull request closed without merging**